cargo-cyclonedx/src/urlencode.rs

```diff
@@ -0,0 +1,85 @@
+//! If you are reading this - buckle up, we are going on an adventure!
+//!
+//! So in the purl spec there is this innocuous example of two valid PURLs:
+//! ```text
+//! pkg:generic/openssl@1.1.10g?download_url=https://openssl.org/source/openssl-1.1.0g.tar.gz&checksum=sha256:de4d501267da
+//! pkg:generic/bitwarderl?vcs_url=git%2Bhttps://git.fsfe.org/dxtr/bitwarderl%40cc55108da32
+//! ```
+//! from <https://github.com/package-url/purl-spec/blob/07d2bdea6d9610b52772866c7ed7859e5566f318/PURL-TYPES.rst#generic>
+//!
+//! Note the `git%2Bhttps` part. The `%2B` is a percent-encoded `+` character, which is necessary, because otherwise
+//! the `+` would be turned into a space when decoding and the original string would not be recoverable.
+//!
+//! I have naively assumed that there is a single, well-defined percent encoding standard.
+//!
+//! In reality The URL spec has numerous different sets of characters that should or should not be URL-encoded.
+//!
+//! This part, `?foo=bar`, is called the "query" in the URL spec:
+//! <https://url.spec.whatwg.org/#concept-url-query>
+//!
+//! And this is what characters are supposed to be URL-encoded there:
+//! <https://url.spec.whatwg.org/#query-percent-encode-set>
+//!
+//! Note the absence of the `+` character in this set!
+//! It is apparently legal to put a + in there, but the parsers I tried convert it into a space!
+//!
+//! There are only two character sets that escape `+`:
+//! 1. <https://url.spec.whatwg.org/#component-percent-encode-set>
+//!    to be used for "components", but the spec NEVER DEFINES WHAT A COMPONENT IS.
+//! 2. <https://url.spec.whatwg.org/#application-x-www-form-urlencoded-percent-encode-set>
+//!    to be used for form submission, so not our case?
```

Comment:
It's a mistake to believe that this is only used for form submission. This same exact encoding is used to set the `query` string of a URL (because that's where form submission data goes when the method is `GET` anyways).

The spec even states that this set will be used [for anything in the `URLSearchParams` class][0], which is used for generating query strings (or body data in the case of `POST` form submission).

[0]: https://url.spec.whatwg.org/#ref-for-application-x-www-form-urlencoded-percent-encode-set%E2%91%A3

That is why, for example, the [whatwg-url][1] package (which is [synchronized][2] with the WHATWG URL spec), encodes these characters when set using the `searchParam` attribute. For example:

```js
import { URL } from 'whatwg-url';
const x = new URL('pkg:generic/bitwarderl');
x.searchParams.set('vcs_url', 'vcs_url=git+https://github.com/rustsec/rustsec.git@6f82509cb43a')
console.log(x.toString());
```

Outputs:

```
pkg:generic/bitwarderl?vcs_url=vcs_url%3Dgit%2Bhttps%3A%2F%2Fgithub.com%2Frustsec%2Frustsec.git%406f82509cb43a
```

[1]: https://www.npmjs.com/package/whatwg-url
[2]: https://github.com/whatwg/url#implementations

It is fine to use characters such as `+` unencoded in the query string, but **not** if you are parsing that query string as search parameters (as PURL does).

Reply:
Even that cannot be used to produce the official example of `?vcs_url=git%2Bhttps://git.fsfe.org/dxtr/bitwarderl%40cc55108da32`, because the `application-x-www-form-urlencoded` set encodes `:` and `/` but they are left unencoded in the official example.

Reply:
I suppose it would be a question to the PURL maintainers to see if they actually intend qualifiers to be interoperable with `URLSearchParams`. If so, mandatory percent encoding of `+` should be mentioned in the spec...

Reply:
@pombredanne is this something you can maybe weigh in on?